### PR TITLE
Deferring the initialization of the GAPIC client

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This PR is meant to reduce the initialization time of Firestore on GCF. GCF uses the Firestore.snapshot_() API to construct the DocumentSnapshot that they return to the user. This call does not require the networking stack, and hence, we can defer this initialization until we make the first request.

In my preliminary benchmark (across 1000 attempts), calling `new Firestore()` without this PR takes 9 milliseconds. With this PR, it takes 0.04 milliseconds. This difference is likely much larger on GCF, as file loading takes much longer than on my beefy MacBook.

This PR is split into two commits. The first commit just renames `Firestore.api.Firestore` to `Firestore._firestoreClient`.

### CODE REVIEW

File: test/write-batch.js [hunk 1]
```diff
@@ -138,104 +148,109 @@ describe('batch support', function() {
...
-                  fieldPath: 'foo',
-                  setToServerValue: 'REQUEST_TIME',
-                },
-              ],
+  beforeEach(() => {
+    return createInstance().then(firestoreClient => {
+      firestore = firestoreClient;
+      writeBatch = firestore.batch();
+
+      firestore._firestoreClient._commit = function(
```
Comment (schmidt-sebastian) [inline test/write-batch.js]: Note that this all formatting.

File: src/index.js [hunk 2]
```diff
@@ -220,22 +228,13 @@ class Firestore extends commonGrpc.Service {
...
-     * @private
-     * @type {object}
-     * @property {FirestoreClient} Firestore The Firestore GAPIC client.
-     */
-    this.api = {
-      Firestore: v1beta1(options).firestoreClient(options),
-    };
-
-    this._referencePath = new ResourcePath('{{projectId}}', '(default)');
+    this._deferredInitialization = null;
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply src/index.js]: I added comments here and renamed `_deferredInitialization`  and `_clientInitialized`. I really wish this was TypeScript and I didn't have to explain in JSDoc what all the types are....

File: src/index.js [hunk 3]
```diff
@@ -220,22 +228,13 @@ class Firestore extends commonGrpc.Service {
...
-    if (options) {
-      if (options.projectId) {
-        validate.isString('options.projectId', options.projectId);
-        this._referencePath = new ResourcePath(options.projectId, '(default)');
-      }
+    if (options && options.projectId) {
+      validate.isString('options.projectId', options.projectId);
+      this._referencePath = new ResourcePath(options.projectId, '(default)');
+    } else {
+      this._referencePath = new ResourcePath('{{projectId}}', '(default)');
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply src/index.js]: Comment added.

**Comment by mikelehen** [comment]
Thanks for the added context.  I'm hopeful / optimistic that this will make a meaningful impact to alleviate the GCF / Functions team's complaints, but without actually measuring and determining where the time is spent, it's extremely easy to make "improvements" that don't actually move the needle... hence my concern. :-)  But this is just food for thought.  Carry on!

**Pull request merged**